vertx-web/src/main/java/io/vertx/ext/web/handler/impl/CorsHandlerImpl.java

```diff
@@ -121,9 +123,15 @@ public void handle(RoutingContext context) {
     HttpServerResponse response = context.response();
     String origin = context.request().headers().get(ORIGIN);
     if (origin == null) {
+      if (allowedOrigin != null) {
+        response.putHeader(VARY, ORIGIN);
+      }
       // Not a CORS request - we don't set any headers and just call the next handler
       context.next();
     } else if (isValidOrigin(origin)) {
+      if (allowedOrigin != null) {
+        response.putHeader(VARY, ORIGIN);
+      }
```

Comment:
Here's the second place you read a request header in order to send a different response depending on it (actually only depending on its presence or absence, but you need a `Vary` header value for that as well). So:

```suggestion
      response.putHeader(VARY, ACCESS_CONTROL_REQUEST_METHOD);
```